src/dom/Value.ts

```diff
@@ -0,0 +1,119 @@
+import {Decimal, IonType} from "../Ion";
+import JSBI from "jsbi";
+import {Constructor} from "./DomValue";
+
+/**
+ * Provides a common interface to work with DOM elements of any Ion type.
+ *
+ * Implementors should return `null` for any method that cannot provide the requested value.
+ */
+export interface Value {
+    /**
+     * Returns the Ion type of this value.
+     */
+    getType(): IonType;
+
+    /**
+     * Returns an array containing any annotations associated with this value.
+     * If this value does not have any annotations, an empty array will be returned.
+     */
+    getAnnotations(): string[];
+
+    /**
+     * Returns true if this Value is a Null and false otherwise.
+     */
+    isNull(): boolean;
+
+    /**
+     * Returns null if the Value is not a Boolean.
+     * Otherwise, returns the boolean value of the Boolean.
+     */
+    booleanValue(): boolean | null;
+
+    /**
+     * Returns null if the Value is not an Integer, Float, or Decimal.
+     * Otherwise, returns the JS number representation of the Value.
+     * In some cases, this method will perform conversions that inherently involve loss of
+     * precision. See bigIntValue() and decimalValue() for lossless alternatives.
+     */
+    numberValue(): number | null;
+
+    /**
+     * Returns null if the Value is not an Integer.
+     * Otherwise, returns a lossless BigInt representation of the Value.
+     */
+    bigIntValue(): JSBI | null;
+
+    /**
+     * Returns null if the Value is not a Decimal.
+     * Otherwise, returns a lossless Decimal representation of the Value.
+     */
+    decimalValue(): Decimal | null;
+
+    /**
+     * Returns null if the Value is not a String or Symbol.
+     * Otherwise, returns a string representation of the Value's text.
+     */
+    stringValue(): string | null;
+
+    /**
+     * Returns null if the Value is not a Timestamp.
+     * Otherwise, returns a Date representation of the Value.
+     */
+    dateValue(): Date | null;
```

Comment:
The methods are named after their return types. `dateValue()` returns a native JS `Date` object. However, I'd meant to offer a `timestampValue(): Timestamp` along side it! I'll go back and add that.

Reply:
> please check the first parameter of the dom-class constructors for consistency.  should others specify `protected readonly`? 

Adding `protected readonly` to a constructor parameter is a shorthand for creating a member field of the same name. We don't need to use it everywhere.

> looks like we're not using a "_" prefix consistently?

Good eye. I like the `protected readonly` shorthand for creating member fields, but disliked that this caused documentation and tooltips to include the `_` prefix so I seem to have mixed the two styles. I'll bite the bullet and use `_` everywhere.

> it's _entirely possible_ i don't understand the beauty of `[...x]` ;), but i expect `x` to be equivalent; am i missing out on the magic?

The `Value` interface's `fieldNames()`, `fields()`, and `values()` methods return `IterableIterator`s. While this is nice for iteration/looping, Javascript doesn't provide `Array`-style convenience methods like `map`, `filter`, and `reduce` for iterators. Likewise, there's no easy way to get their length.

The idiomatic JS solution to this problem is to use the spread operator to create an intermediate array you can perform these operations on. (`[...v.values()]`)

I had originally written `fieldNames()`, `fields()`, and `values()` to return arrays instead of iterators, but ran into a snag: `values()` is already a method on the `Array` type and it returns an `IterableIterator<any>`. I wasn't able to override it without also returning an iterator.

If you can suggest another name for `values()`, I can switch the return types to Arrays, probably making everyone's lives easier. `children()`?

Reply:
> I like the protected readonly shorthand for creating member fields, but disliked that this caused documentation and tooltips to include the _ prefix so I seem to have mixed the two styles. I'll bite the bullet and use _ everywhere.

If `_` is showing up in the documentation, I'd prefer we prevent that we some `@ignore` incantation, or fall back to not using the shorthand.  The documentation should correspond to the public API.

---

> If you can suggest another name for values(), I can switch the return types to Arrays, probably making everyone's lives easier. children()?

I think you're proceeding with `elements()`